javaslang/src/main/java/javaslang/collection/BitSet.java

```diff
@@ -0,0 +1,1024 @@
+        final Comparator<? super T> component1Comparator = comparator();
+        final Comparator<Tuple2<T, Long>> tuple2Comparator = (t1, t2) -> component1Comparator.compare(t1._1, t2._1);
+        return TreeSet.ofAll(tuple2Comparator, iterator().zipWithIndex());
+    }
+}
+
+interface BitSetModule {
+
+    int ADDRESS_BITS_PER_WORD = 6;
+    int BITS_PER_WORD = 64;
+
+    abstract class AbstractBitSet<T> implements BitSet<T> {
+
+        private static final long serialVersionUID = 1L;
+
+        final Function1<Integer, T> fromInt;
+        final Function1<T, Integer> toInt;
+
+        AbstractBitSet(Function1<Integer, T> fromInt, Function1<T, Integer> toInt) {
+            this.fromInt = fromInt;
+            this.toInt = toInt;
+        }
+
+        abstract int getWordsNum();
+
+        abstract long[] copyExpand(int wordsNum);
+
+        abstract long getWord(int index);
+
+        BitSet<T> createEmpty() {
+            return new BitSet1<>(fromInt, toInt, 0L);
+        }
+
+        BitSet<T> createFromAll(Iterable<? extends T> values) {
+            return createEmpty().addAll(values);
+        }
+
+        BitSet<T> fromBitMaskNoCopy(long[] elements) {
+            final int len = elements.length;
+            if (len == 0) {
+                return createEmpty();
+            }
+            if (len == 1) {
+                return new BitSet1<>(fromInt, toInt, elements[0]);
+            }
+            if (len == 2) {
+                return new BitSet2<>(fromInt, toInt, elements[0], elements[1]);
+            }
+            return new BitSetN<>(fromInt, toInt, elements);
+        }
+
+        private void setElement(long[] words, int element) {
+            final int index = element >> ADDRESS_BITS_PER_WORD;
+            words[index] |= (1L << element);
+        }
+
+        private void unsetElement(long[] words, int element) {
+            final int index = element >> ADDRESS_BITS_PER_WORD;
+            words[index] &= ~(1L << element);
+        }
+
+        long[] shrink(long[] elements) {
+            int newlen = elements.length;
+            while (newlen > 0 && elements[newlen - 1] == 0) {
+                newlen--;
+            }
+            long[] newelems = new long[newlen];
+            System.arraycopy(elements, 0, newelems, 0, newlen);
+            return newelems;
+        }
+
+        @Override
+        public BitSet<T> add(T t) {
+            if (contains(t)) {
+                return this;
+            } else {
+                final int element = toInt.apply(t);
+                final long[] copy = copyExpand(1 + (element >> ADDRESS_BITS_PER_WORD));
+                setElement(copy, element);
+                return fromBitMaskNoCopy(copy);
+            }
+        }
+
+        @Override
+        public BitSet<T> distinctBy(Comparator<? super T> comparator) {
+            Objects.requireNonNull(comparator, "comparator is null");
+            return createFromAll(iterator().distinctBy(comparator));
```

Comment:
could we return the original in case of equality? Or it's unreasonable in this case?